File: python/problem1505-2.py
class Solution:
    
    def minInteger(self, num: str, k: int) -> str:
        n = len(num)
        digits = [str(i) for i in range(10)]
        # 线段树上0代表未被换到前面去，1代表被换了
        # 所以getSum得到的就是当前位置前面被换过了的数量
        # 等价于未处理字符串的开头位置
        segmentTree = SegmentTree(n)
        # 记录0~9各个数字的位置
        position = [[] for _ in range(10)]
        for i in range(n - 1, -1, -1):
            position[int(num[i])].append(i)
        res = []
        i = 0
        while i < n:
            digit = int(num[i])
            if not position[digit] or position[digit][-1] > i:
                # 已经被交换过的了，忽略掉
                i += 1
                continue
            for j in range(digit + 1):
                if not position[j]:
                    continue
                # 在num中的索引
                originIdx = position[j][-1]
                # 找到一个比num[i]小的了，但还需要看距离是否够
                count = segmentTree.getSum(originIdx + 1)
                # 实际需要多少次的交换
                needCount = originIdx - count
                if needCount <= k:
                    k -= needCount
                    position[j].pop()
                    res.append(str(j))
                    segmentTree.update(originIdx + 1, 1)
                    break
        return ''.join(res)


class SegmentTree:
    def __init__(self, n):
        self.n = n
        self.tree = [0] * (n + 1)
    
    def lowbit(self, x):
        return x & (-x)
    
    def update(self, i, x):
        while i <= self.n:
            self.tree[i] += x
            i += self.lowbit(i)

    def getSum(self, i):
        res = 0
        while i > 0:
            res += self.tree[i]
            i -= self.lowbit(i)
        return res

File: Java/problem1505-2.java
class Solution {
    public String minInteger(String num, int k) {
        char[] nums = num.toCharArray();
        int n = nums.length;
        SegmentTree segmentTree = new SegmentTree(n);
        LinkedList[] position = new LinkedList[10];
        for (int i = 0; i < 10; ++i) {
            position[i] = new LinkedList<Integer>();
        }
        for (int i = n - 1; i >= 0; --i) {
            position[nums[i] - '0'].addLast(i);
        }
        int i = 0;
        StringBuilder sb = new StringBuilder();
        while (i < n) {
            int digit = nums[i] - '0';
            if (position[digit].size() == 0 || (int)(position[digit].getLast()) > i) {
                ++i;
                continue;
            }
            for (int j = 0; j <= digit; ++j) {
                if (position[j].size() == 0) {
                    continue;
                }
                int originIdx = (int)position[j].getLast();
                int count = originIdx - segmentTree.getSum(originIdx + 1);
                if (count <= k) {
                    k -= count;
                    sb.append(nums[originIdx]);
                    position[j].removeLast();
                    segmentTree.update(originIdx + 1, 1);
                    break;
                }
            }
        }
        return sb.toString();
    }
}

class SegmentTree {
    int n;
    int[] tree;

    public SegmentTree(int val) {
        n = val;
        tree = new int[val + 1];
    }

    public int lowbit(int x) {
        return x & (-x);
    }

    public void update(int idx, int x) {
        while (idx <= n) {
            tree[idx] += x;
            idx += lowbit(idx);
        }
    }

    public int getSum(int idx) {
        int res = 0;
        while (idx > 0) {
            res += tree[idx];
            idx -= lowbit(idx);
        }
        return res;
    }
}
